src/Std/Data/DHashMap/Internal/List/Associative.lean

```diff
@@ -1827,4 +1903,1213 @@ theorem eraseKey_append_of_containsKey_right_eq_false [BEq α] {l l' : List ((a
+  rw [← getValue?_eq_some_getValue]
+  rw [getValue?_insertList_of_contains_eq_false]
+  exact h'
+
+theorem getValue_insertList_of_mem [BEq α] [EquivBEq α]
+    {l toInsert : List ((_ : α) × β)}
+    {k k' : α} (k_beq : k == k') {v : β}
+    (distinct : toInsert.Pairwise (fun a b => (a.1 == b.1) = false))
+    (distinct2 : DistinctKeys l)
+    (mem : ⟨k, v⟩ ∈ toInsert)
+    {h} :
+    getValue k' (insertList l toInsert) h = v := by
+  rw [← Option.some_inj]
+  rw [← getValue?_eq_some_getValue]
+  rw [getValue?_insertList_of_mem]
+  . exact k_beq
+  . exact distinct
+  . exact distinct2
+  . exact mem
+
+theorem getValue!_insertList_of_contains_eq_false [BEq α] [PartialEquivBEq α] [Inhabited β]
+    {l toInsert : List ((_ : α) × β)} {k : α}
+    (h : (toInsert.map Sigma.fst).contains k = false) :
+    getValue! k (insertList l toInsert) = getValue! k l := by
+  simp only [getValue!_eq_getValue?]
+  rw [getValue?_insertList_of_contains_eq_false]
+  exact h
+
+theorem getValue!_insertList_of_mem [BEq α] [EquivBEq α] [Inhabited β]
+    {l toInsert : List ((_ : α) × β)} {k k' : α} {v: β} (k_beq : k == k')
+    (distinct : toInsert.Pairwise (fun a b => (a.1 == b.1) = false))
+    (distinct2 : DistinctKeys l)
+    (mem : ⟨k, v⟩ ∈ toInsert):
+    getValue! k' (insertList l toInsert) = v := by
+  simp only [getValue!_eq_getValue?]
+  rw [getValue?_insertList_of_mem]
+  · rw [Option.get!_some]
+  · exact k_beq
+  · exact distinct
+  . exact distinct2
+  . exact mem
+
+theorem getValueD_insertList_of_contains_eq_false [BEq α] [PartialEquivBEq α]
+    {l toInsert : List ((_ : α) × β)} {k : α} {fallback : β}
+    (h : (toInsert.map Sigma.fst).contains k = false) :
+    getValueD k (insertList l toInsert) fallback = getValueD k l fallback := by
+  simp only [getValueD_eq_getValue?]
+  rw [getValue?_insertList_of_contains_eq_false]
+  exact h
+
+theorem getValueD_insertList_of_mem [BEq α] [EquivBEq α]
+    {l toInsert : List ((_ : α) × β)} {k k' : α} {v fallback: β} (k_beq : k == k')
+    (distinct : toInsert.Pairwise (fun a b => (a.1 == b.1) = false))
+    (distinct2 : DistinctKeys l)
+    (mem : ⟨k, v⟩ ∈ toInsert):
+    getValueD k' (insertList l toInsert) fallback= v := by
+  simp only [getValueD_eq_getValue?]
+  rw [getValue?_insertList_of_mem]
+  · rw [Option.getD_some]
+  · exact k_beq
+  · exact distinct
+  . exact distinct2
+  . exact mem
+
+
+-- FIXME: This already exists in mathlib...
+def Prod.toSigma (p : α × β) : ((_ : α) × β) := ⟨p.fst, p.snd⟩
+
+def List.toSigma (l : List (α × β)) : List ((_ : α) × β) := l.map Prod.toSigma
+
+@[simp]
+theorem sigma_fst_after_prod_toSigma_eq_prod_fst :
+    Sigma.fst ∘ Prod.toSigma = @Prod.fst α β := by
+  apply funext
+  simp [Prod.toSigma]
+
+-- TODO: maybe insertList_prod would be a better name?
+/-- Internal implementation detail of the hash map -/
+def insertListConst [BEq α] (l: List ((_ : α) × β)) (toInsert: List (α × β)) : List ((_ : α) × β) :=
+  insertList l toInsert.toSigma
```

Comment:
The machinery around `insertListConst` mostly comes down to mappings between `Prod` and `Sigma`. 
Maybe it would be appreciated to at least have `Prod.toSigma` and somethings like `sigma_fst_after_prod_toSigma_eq_prod_fst` at a more appropriate place in the codebase. (Note that mathlib already features `Prod.toSigma`.)
What would be your suggestion to go about this?
I felt like this was making the work around `insertListConst` a lot easier with fewer auxiliary theorems. I would also be happy with only having this in the scope of this file though :)

Reply:
Yes, I agree that having `Prod.toSigma` makes a lot of sense. In the interest of keeping this PR simple, I would suggest keeping the definition in this file for now, with the goal of "promoting" it to a public API in the future. I'm not completely sold on `List.toSigma`, I think the only thing that having this definition does is make you unfold it everywhere. So I suggest dropping it and just defining `insertListCons` as ` insertList l (toInsert.map Prod.toSigma)`.

Reply:
Agreed, `List.toSigma` is not really needed :)